src/d3a/models/strategy/external_strategies/pv.py

```diff
@@ -374,52 +371,76 @@ def __init__(
                          update_interval=update_interval,
                          energy_rate_decrease_per_update=energy_rate_decrease_per_update,
                          use_market_maker_rate=use_market_maker_rate)
-        self.energy_forecast_buffer_Wh = energy_forecast_Wh
+
+        # Buffers for energy forecast and measurement values,
+        # that have been sent by the d3a-api-client in the duration of one market slot
+        self.energy_forecast_buffer = None  # Dict[DateTime, float]
+        self.energy_measurement_buffer = None  # Dict[DateTime, float]
 
     @property
     def channel_dict(self):
+        """Extend channel_dict property with forecast related channels."""
         return {**super().channel_dict,
-                f'{self.channel_prefix}/set_energy_forecast': self._set_energy_forecast}
+                f"{self.channel_prefix}/set_energy_forecast": self._set_energy_forecast,
+                f"{self.channel_prefix}/set_energy_measurement": self._set_energy_measurement}
 
     def event_tick(self):
         """Set the energy forecast using pending requests. Extends super implementation.
 
         This method is triggered by the TICK event.
         """
-        # Need to repeat he pending request parsing in order to handle energy forecasts
+        # Need to repeat the pending request parsing in order to handle energy forecasts
         # from the MQTT subscriber (non-connected admin)
         for req in self.pending_requests:
             if req.request_type == "set_energy_forecast":
                 self._set_energy_forecast_impl(req.arguments, req.response_channel)
+            elif req.request_type == "set_energy_measurement":
+                self._set_energy_measurement_impl(req.arguments, req.response_channel)
 
         self.pending_requests = deque(
             req for req in self.pending_requests
-            if req.request_type not in "set_energy_forecast")
+            if req.request_type not in ["set_energy_forecast", "set_energy_measurement"])
 
         super().event_tick()
 
     def _incoming_commands_callback_selection(self, req):
+        """Map commands to callbacks for forecast and measurement reading."""
         if req.request_type == "set_energy_forecast":
             self._set_energy_forecast_impl(req.arguments, req.response_channel)
+        elif req.request_type == "set_energy_measurement":
+            self._set_energy_measurement_impl(req.arguments, req.response_channel)
         else:
             super()._incoming_commands_callback_selection(req)
```

Comment:
Couldn't this method be used in `event_tick()` to decide which callback to call?

Reply:
It is used in the `event_tick()`, have a look into the `PVExternalMixin`.

Reply:
Sorry, I meant in `PVForecastExternalStrategy`.

Reply:
Sorry, I dont follow. Can you please explain some more? Thanks!

Reply:
The `event_tick()` method in `PVForecastExternalStrategy` doesn't call `_incoming_commands_callback_selection()`, but it looks like it duplicates its code.

Reply:
`PVForecastExternalStrategy._incoming_commands_callback_selection` extends the usage of `PVExternalMixin._incoming_commands_callback_selection`. We need to also call `PVExternalMixin.event_tick` (and the `PVExternalMixin._incoming_commands_callback_selection`) in the `PVForecastExternalStrategy.event_tick`.
 The duplication you see was added by Spyros, which also added a comment where he mentions that this is needed to make it work for the MQTT subscriber. I am willing to add a ticket that looks into the need for this duplication (which is probably not needed any more because of the fix i added [here](https://github.com/gridsingularity/d3a/pull/1200#discussion_r690172774). What do you think?

Reply:
I think that the comment is referring to the repetition of that operation, rather than to the duplication of the code. What I meant is that those lines could maybe be replaced by
```python
self._incoming_commands_callback_selection(req)
```
Anyway, this is not essential and I just noticed I should have prepended it with "Nit".

Reply:
Sorry, still do not get it. We need to take care of the req.type's `set_energy_forecast` and `set_energy_measurement` here. for all other req.type's  the `super()._incoming_commands_callback_selection(req)` will take care of . Right?

Reply:
Maybe we're getting confused :D 
In `PVForecastExternalStrategy` we have:
```python
def event_tick(self):
    # ...
    for req in self.pending_requests:
        if req.request_type == "set_energy_forecast":
            self._set_energy_forecast_impl(req.arguments, req.response_channel)
        elif req.request_type == "set_energy_measurement":
            self._set_energy_measurement_impl(req.arguments, req.response_channel)
    # ...
    super().event_tick()
```
But it could be simplified to be:
```python
def event_tick(self):
    # ...
    for req in self.pending_requests:
        self._incoming_commands_callback_selection(req)
    # ...
    super().event_tick()
```
Because `_incoming_commands_callback_selection(req)` is defined inside `PVForecastExternalStrategy` and it already checks `set_energy_forecast` and `set_energy_measurement`. I hope this makes sense.

Reply:
@fievelk Please have a look at [this inline comment](https://github.com/gridsingularity/d3a/pull/1200/files#diff-44fccfc06172fd26d724ae1e5828f14d8e0a6e4d9dfa259090365298a9ba1410L389). @spyrostz added the duplication for a reason, that he maybe can explain better. I fear that this was a fix for this bug: https://github.com/gridsingularity/d3a/pull/1200#discussion_r690172774

Reply:
The main problem here is that the event_tick of the PVExternalMixin base class is checking the `self.connected` member and does not trigger the call to `_incoming_commands_callback_selection` if no user has registered / connected to the PV before (that's what the connected flag controls, it essentially checks if a user has registered to this device before accepting commands). Unfortunately this whole model breaks for the MQTT subscriber, where we need to send measurements directly to the device without registering first, using an admin account and authenticating only in d3a-web / client api service. This is what the comment describes, that specifically for this case we have to circumvent the normal flow and deliberately accept energy forecasts (and measurements IMO, since we are in the topic) without relying on the super().event_tick() call.

Reply:
@spyrostz Thanks for the explanation about the comment and this part of the code, which is now clearer to me 👍

Still, I don't think this is connected to what we were discussing earlier. The refactoring I proposed doesn't change the functionality of the code. The comment can stay there, the parsing will happen as before. It's just that we would re-use code (already defined in `PVForecastExternalStrategy._incoming_commands_callback_selection()`) instead of duplicating it. If this doesn't make sense, please ignore my suggestion: it might be that I'm not seeing something here and this was just a nit :)

Reply:
@fievelk no worries at all, I did not read carefully enough your comment and did not figure out that this was your exact suggestion, thanks for clarifying it! 
There is one tricky cornercase with your suggestion: `_incoming_commands_callback_selection` also uses super() https://github.com/gridsingularity/d3a/blob/96a69bf0ba8029991a87a0e397906105514feecb/src/d3a/models/strategy/external_strategies/pv.py#L405. This means that if we call this function from the PVForecastExternalStrategy event_tick method, all the commands will be processed immediately (including offer, bid ...) which is not the desirable outcome here (we only want the forecast and measurement commands to circumvent the normal process, not every command). Of course, the `_incoming_commands_callback_selection` method could be refactored to avoid this (e.g. by moving the if blocks for the forecast and measurement to their own method and calling this method twice), I am not sure if it will make the code more complex/unreadable than it already is.

Reply:
Thanks! I did not think about the fact that processing the other commands wasn't desirable in this case! Thanks for the patience :D